fix(ui): add support for Ollama model tags in the UI

sbx0r opened the pull request:
This PR /resolves #307 issue. It adds ability to specify custom model tags when creating or editing Ollama model configurations in the UI

Changes:
- Added a "Model Tag" input field that appears only when Ollama provider is selected
- Default tag value is set to "latest" when not specified
- Automatically includes the tag in the model name when it's different from "latest"
- Sends the model name to the API in the format model:tag when a custom tag is specified
- Stores the tag value in the model parameters for reference
- Updates the auto-generated model resource name to include the tag for better identification
- The tag input helps users specify which version of an Ollama model they want to use, making it easier to work with different versions of the same model.

The tag input helps users specify which version of an Ollama model they want to use, making it easier to work with different versions of the same model.

> [!IMPORTANT]
> This change currently **prevents updating** the model tag on existing ModelProvider resources.
>
> **Here’s why:**
> When using the automatic name generator, the resulting ModelProvider name follows the convention
> {provider}-{model}-{tag} — for example, ollama-starcoder2-3b.
> If the tag is changed without updating the resource name, it may lead to confusion — e.g., a model named ollama-deepseek-r1-7b might actually point to 70b, or vice versa.
>
> To ensure stability, the controller must either:
>   - guarantee that all dependent Teams/Agents will use the updated model, or
>   - prohibit such updates altogether.
>
> This needs to be discussed further.

--- 

After this PR merges the 'Create New Model' card looks like this:
- [non-ollama]
![image](https://github.com/user-attachments/assets/f55bd7f8-01e9-4e08-9b06-9d6dcb2befa6)
- [ollama with latest tag]
![image](https://github.com/user-attachments/assets/2b7759e9-cbe0-4b6f-a7ae-14a6585056ff)
- [ollama with non-latest tag]
![image](https://github.com/user-attachments/assets/913aecd6-5848-45ce-bf9e-e19e0189b037)

Resulting model will look like this:

- [ui]
![image](https://github.com/user-attachments/assets/1f1f0237-6df8-4360-b261-17eef71d5d8b)
- [resulting k8s manifest]
  ```yaml
   apiVersion: kagent.dev/v1alpha1
   kind: ModelConfig
   metadata:
     name: ollama-dolphin-mistral-70b
     namespace: kagent
   spec:
     apiKeySecretKey: ""
     apiKeySecretRef: ""
     model: dolphin-mistral:70b
     ollama: {}
     provider: Ollama
  ```

peterj commented on ui/src/app/models/new/page.tsx:
```diff
@@ -366,13 +384,27 @@ function ModelPageContent() {
+        const tagSuffix = `-${modelTag.trim()}`;
+        finalModelName = finalModelName + tagSuffix;
+
+        if (!isResourceNameValid(finalModelName)) {
```
recently I added `createRFC1123ValidName` -- perhaps use that to create the name? (so you don't have to check whether it's valid and use a regex).

sbx0r replied:
current implementation of createRFC1123ValidName have this logic

```ts
  //...

  // Final validation check - though sanitization should make it valid
  if (!isResourceNameValid(combined)) {
      console.warn(`Generated name '${combined}' is still not RFC 1123 valid after sanitization.`);
      // Returning potentially invalid name, caller should ideally re-validate
      // Or return "" to indicate failure? Returning the attempt might be more informative.
      return combined;
  }

  //...
```

Could you tell why is it there?

peterj replied:
good point -- I think that was just a check in case it still couldn't create the valid name; I think we could keep this as is and see if anyone even runs into it.

sbx0r replied:
alright. should I validate again after the function call or the console.warn will be enough?

peterj replied:
I think warn should be enough -- and we'll see in which scenarios users run into this (if ever)

peterj commented on ui/src/app/models/new/page.tsx:
```diff
@@ -264,14 +278,18 @@ function ModelPageContent() {
-        const baseName = `${providerKey}-${modelName}`.toLowerCase();
+        let baseName = `${providerKey}-${modelName}`.toLowerCase();
+        const isOllama = selectedProvider?.type === "Ollama";
+        if (isOllama && modelTag && modelTag !== 'latest') {
```
nit: create a constant outside of the component for `latest` (E.g. `const LATEST_TAG="latest"`)

peterj commented on ui/src/app/models/new/page.tsx:
```diff
@@ -114,6 +114,7 @@ function ModelPageContent() {
   const [providerModelsData, setProviderModelsData] = useState<ProviderModelsResponse | null>(null);
   const [selectedCombinedModel, setSelectedCombinedModel] = useState<string | undefined>(undefined);
   const [selectedModelSupportsFunctionCalling, setSelectedModelSupportsFunctionCalling] = useState<boolean | null>(null);
```
perhaps you can do the change in the wizard as well as a separate PR (doesn't have to be part of this one). Thanks a ton for doing this!

sbx0r replied:
My pleasure!
I'm flexible. If you'd like this change quickly, I can open a follow-up issue an PR.
LMK what works best for you.

peterj replied:
it's up to you -- if you don't mind adding it as part of this PR that would be good. but if not, a separate pr/issue is great as well.

sbx0r replied:
I will add it to this PR then :)

peterj commented:
another issue I noticed -- looks like the "0.5b" tag is applied twice to the name (I think this will be probably resolved once you switch to the method to create valid name).

<img width="756" alt="Screenshot 2025-05-04 at 2 43 35 PM" src="https://github.com/user-attachments/assets/100af725-1c2d-46e4-a21f-a93d8ae220e2" />

peterj commented:
also, check that the editing works - it looks like the provider is not filled out and the tag is readonly
<img width="1330" alt="Screenshot 2025-05-04 at 2 50 06 PM" src="https://github.com/user-attachments/assets/9a8f427c-e72c-4a78-ad6a-5a0de42c560e" />
as well -- right now, it

sbx0r commented:
> another issue I noticed -- looks like the "0.5b" tag is applied twice to the name (I think this will be probably resolved once you switch to the method to create valid name).
> 
> <img alt="Screenshot 2025-05-04 at 2 43 35 PM" width="756" src="https://private-user-images.githubusercontent.com/11080940/440218598-100af725-1c2d-46e4-a21f-a93d8ae220e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYzOTU0NTEsIm5iZiI6MTc0NjM5NTE1MSwicGF0aCI6Ii8xMTA4MDk0MC80NDAyMTg1OTgtMTAwYWY3MjUtMWMyZC00NmU0LWEyMWYtYTkzZDhhZTIyMGUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA0VDIxNDU1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZjMjJkNzc5ZWQ1NDZlNzY5YmYyNzViMWNjYzEzYWU4MWYzM2JjM2YzYjY5NzBjNTdmMTg5NDMxMDAxOTIxNDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AB8zCo7cZsJ3rfex108zLEFj_n22SxCAQnAdCNjwN7U">

Thanks for checking it out!
I’ll use your validator and run it against the edge case you spotted.

sbx0r commented:
> also, check that the editing works - it looks like the provider is not filled out and the tag is readonly <img alt="Screenshot 2025-05-04 at 2 50 06 PM" width="1330" src="https://private-user-images.githubusercontent.com/11080940/440218923-9a8f427c-e72c-4a78-ad6a-5a0de42c560e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYzOTU3NzcsIm5iZiI6MTc0NjM5NTQ3NywicGF0aCI6Ii8xMTA4MDk0MC80NDAyMTg5MjMtOWE4ZjQyN2MtZTcyYy00YTc4LWFkNmEtNWEwZGU0MmM1NjBlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA0VDIxNTExN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiNjBiYjQ0NDhhOTJmYzYwODM4NjE4NjI0Mjk4ODRlZmY2ZmY3NmE5OTA3OWVkYjA4YjZhOTM4NjZjMGQxYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.49muooGaezmweteH4dMJKJnH_lB54cO1sYSV9qO87kQ"> as well -- right now, it

Yup. I found the issue with empty provider but didn't want to add noise to the PR. I can fix it here if that's okay with you.
For the model tag lock - I've some thoughts. Please check the PR description (see the 'Important' section)

peterj commented:
> > also, check that the editing works - it looks like the provider is not filled out and the tag is readonly <img alt="Screenshot 2025-05-04 at 2 50 06 PM" width="1330" src="https://private-user-images.githubusercontent.com/11080940/440218923-9a8f427c-e72c-4a78-ad6a-5a0de42c560e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYzOTU3NzcsIm5iZiI6MTc0NjM5NTQ3NywicGF0aCI6Ii8xMTA4MDk0MC80NDAyMTg5MjMtOWE4ZjQyN2MtZTcyYy00YTc4LWFkNmEtNWEwZGU0MmM1NjBlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA0VDIxNTExN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiNjBiYjQ0NDhhOTJmYzYwODM4NjE4NjI0Mjk4ODRlZmY2ZmY3NmE5OTA3OWVkYjA4YjZhOTM4NjZjMGQxYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.49muooGaezmweteH4dMJKJnH_lB54cO1sYSV9qO87kQ"> as well -- right now, it
> 
> Yup. I found the issue with empty provider but didn't want to add noise to the PR. I can fix it here if that's okay with you. For the model tag lock - I've some thoughts. Please check the PR description (see the 'Important' section)

It would be a pain to go and update all agents when renaming the model; but I don't want to prevent people from updating the models once they created them. Perhaps we can only allow editing the models parameters (e.g. API key, baseUrl, etc.) but not the actual model name/tag.

sbx0r commented:
All is done.
I also added minor UX improvement to the wizard (set dynamic `model config name` if not entered manually by user) and fixed provider icons to make 'em visible in both - light and dark modes

peterj commented on helm/kagent/templates/deployment.yaml:
```diff
@@ -50,6 +50,10 @@ spec:
           {{- end }}
+          ports:
+            - name: http
+              containerPort: {{ .Values.service.ports.controller.targetPort }}
```
@EItanya you fine with exposing this (mostly for testing purposes)?

EItanya replied:
Yea I don't see why not, seems good to have anyway

peterj commented:
@sbx0r  Looks like there are some linting errors in the UI (you can run `npm run build` locally to see the issues/fix them:))

sbx0r commented:
> @sbx0r Looks like there are some linting errors in the UI (you can run `npm run build` locally to see the issues/fix them:))
Thanks!
I'm working on it.
Should be ready in a minute

peterj commented:
@sbx0r , looks like that one issue I mentioned earlier is still there -- if I use the tag "0.5b", in the creation page, the name of the resource looks right (e.g. `ollama-qwen2-5-0-5b`), however, when the resource is actually created, it creates "ollama-qwen2-5-0-5b-0.5b".

sbx0r commented:
> @sbx0r , looks like that one issue I mentioned earlier is still there -- if I use the tag "0.5b", in the creation page, the name of the resource looks right (e.g. `ollama-qwen2-5-0-5b`), however, when the resource is actually created, it creates "ollama-qwen2-5-0-5b-0.5b".

hmmm.....
Looks like an issue related to the OS/Browser/etc.

It works like a charm on my machine
![image](https://github.com/user-attachments/assets/4351c64c-5b27-493a-871a-4a70a422e531)

![image](https://github.com/user-attachments/assets/3cd38859-85e2-4b88-bd28-b75d6f9b5488)

peterj commented:
you're completely right; this was me -- the GH pull request extension doesn't actually refresh the PR so I had old code -- anyway. it works -- let's wait for the CI and it's good to go!

sbx0r commented:
Great to hear (that it's not a platform issue LoL 😄)!
